### PR TITLE
[FIX] account: analytic accounts can be deleted even if used

### PR DESCRIPTION
**Pull request by momegahed**
Steps to reproduce:

- create analytic account test
- create a draft invoice with lines that uses test
- delete test
- now try to edit the lines previously created
- infinite error

Fix:
creating a `ondelete` hook to more forcibly apply the constraint to contain both draft and posted invoices. the current behavior prevents the deletion of analytic accounts only if used in posted invoices. this happens because the action of posting the invoice creates the `account.analytic.line`s which then have a foreign key constraint with `account.analytic.account` due to m2o relation.

OPW-3235700

### CODE REVIEW

File: addons/account/models/account_analytic_account.py [hunk 1]
```diff
@@ -100,3 +101,18 @@ def action_view_vendor_bill(self):
...
         }
         return result
+
+    @api.ondelete(at_uninstall = False)
+    def _unlink_except_account_set_on_account_move_line(self):
+        query = self.env['account.move.line']._search([])
+        query.order = None
+        query.add_where(
+            'account_move_line.analytic_distribution ?| %s',
+            [[str(account_id) for account_id in self.ids]],
```
Comment (nle-odoo) [inline addons/account/models/account_analytic_account.py:111]: this seemed a little dangerous, but this seems correctly escaped in the resulting query:

```sql
self = self.env['account.analytic.account'].browse(["h'ell[]b\\'obby table\"o"])
# results in
SELECT "account_move_line".id FROM "account_move_line" WHERE account_move_line.analytic_distribution ?| ARRAY['h''ell[]bobby table"o'] ORDER BY  "account_move_line"."date" DESC,"account_move_line"."move_name" DESC,"account_move_line"."id"
```

**Review by nle-odoo — APPROVED** [review]
looks good to me

as discussed, the query is checking all account.move.line().analytic_distribution, but I'm not sure how frequent deleting `account.analytic.account` is

File: addons/account/models/account_analytic_account.py [hunk 2]
```diff
@@ -100,3 +101,18 @@ def action_view_vendor_bill(self):
...
+        )
+        query_string, query_param = query.select('DISTINCT account_move_line.move_name')
+        self._cr.execute(query_string, query_param)
+        move_names = [str(line.get('move_name', '')) for line in self._cr.dictfetchall()]
+        if any(move_names):
+            raise UserError(_('You cannot remove the accounts "%s" because they are used in lines in invoices "%s"'
```
Comment (nle-odoo) [inline addons/account/models/account_analytic_account.py:117]: new terms needs to be added to i18n/account.pot before merge

**Comment by pivi-odoo** [comment]
I would also suggest adding a `limit=1` on the `_search`, because in large databases, `account.move.line` tends to grow quite rapidly, so it is useless to report the names of the problematic invoices. (when invoices are posted, we get an fkey contrains error which doesn't specify which record is problematic, so having another UX for draft invoices isn't necessary. Also what happens in your case if there are let's say 500 `account.move` that have lines that use the AA, the error message box will flood the screen?)
Also if the limit is 1 and we follow this philosophy, you can drop the select of the `move_name`, it's not necessary, since we test for existence, we don't care about who is the culprit.
This is just my opinion.

File: addons/account/models/account_analytic_account.py [hunk 3]
```diff
@@ -100,3 +101,18 @@ def action_view_vendor_bill(self):
...
+    def _unlink_except_account_set_on_account_move_line(self):
+        query = self.env['account.move.line']._search([])
+        query.order = None
+        query.add_where(
+            'account_move_line.analytic_distribution ?| %s',
+            [[str(account_id) for account_id in self.ids]],
+        )
+        query_string, query_param = query.select('DISTINCT account_move_line.move_name')
+        self._cr.execute(query_string, query_param)
+        move_names = [str(line.get('move_name', '')) for line in self._cr.dictfetchall()]
```
[the remark this answers is not in the document]
Comment (pivi-odoo) [reply addons/account/models/account_analytic_account.py:115]: Sadly this doesn't remove duplicate `move_name`?

**Comment by nle-odoo** [comment]
> so having another UX for draft invoices isn't necessary.

if this is only for draft invoice, an improvement could also be to just limit this for draft invoice (I would guess that in normal use case, proportion would be like 95% non-draft vs 5% draft), eg. filtering on parent_state == 'draft')

**Comment by h4818** [comment]
This doesn't cover sales, purchases, and anything else that uses the analytic distribution
@gawa-odoo

File: addons/account/models/account_analytic_account.py [hunk 4]
```diff
@@ -100,3 +101,18 @@ def action_view_vendor_bill(self):
...
+        if any(move_names):
+            raise UserError(_('You cannot remove the accounts "%s" because they are used in lines in invoices "%s"'
+                              , ' ,'.join(self.mapped('name')), ' ,'.join(move_names)))
```
Comment (h4818) [inline addons/account/models/account_analytic_account.py:118]: This could be a really long list, is it necessary to list the invoices

**Comment by h4818** [comment]
Perhaps this could work? https://github.com/odoo/odoo/pull/117762

File: addons/account/models/account_analytic_account.py (hunk 4 above)
Comment (h4818) [reply addons/account/models/account_analytic_account.py:118]: Indeed, it's as frustrating as:
> The operation cannot be completed: another model requires the record being deleted. If possible, archive it instead.
> Model: Analytic Line (account.analytic.line)
> Constraint: account_analytic_line_account_id_fkey

**Comment by Levizar** [comment]
As discussed, I'm closing the PR as the main issue has been solved here: https://github.com/odoo/odoo/pull/117762
We won't add the constraint at the moment as it involves some complexities that @h4818 pointed out.
We would need to do that check in a generic way.

**Pull request closed without merging**